azure/src/main/scala/quasar/physical/blobstore/azure/AzureDatasource.scala

```diff
@@ -18,26 +18,44 @@ package quasar.physical
 package blobstore
 package azure
 
+import slamdata.Predef._
+
 import quasar.api.datasource.DatasourceType
 import quasar.blobstore.azure.{converters => _, _}
+import quasar.blobstore.services._
 import quasar.connector.MonadResourceErr
 
-import cats.effect.{ConcurrentEffect, ContextShift}
+import cats.effect.{ConcurrentEffect, ContextShift, Timer}
+import cats.effect.concurrent.Ref
+import cats.syntax.flatMap._
 import cats.syntax.functor._
+import com.azure.storage.blob.BlobContainerAsyncClient
 import com.azure.storage.blob.models.BlobProperties
 
 object AzureDatasource {
   val dsType: DatasourceType = DatasourceType("azure", 1L)
 
-  def mk[F[_]: ConcurrentEffect: ContextShift: MonadResourceErr](cfg: AzureConfig)
+  def withRefresh[F[_]: ConcurrentEffect, A](
```

Comment:
We could avoid the changes to `BlobstoreDatasource` if we embed the client access and refresh action in the existing effects
```scala
def withRefresh[F[_]: Monad, A, B](
    refClient: Ref[F, Expires[BlobContainerAsyncClient]],
    refreshToken: F[Unit],
    f : BlobContainerAsyncClient => Kleisli[F, A, B])
    : Kleisli[F, A, B] =
```
Also, I think this function can be `private`.

Reply:
Right yeah, I first had an implementation that didn't change `BlobstoreDatasource` but it refreshed at the wrong time (I think at least, not tested). I didn't have the refresh inside the `Kleisli`-"step" in that impl though, so will try that now. 👍